### DISCUSSION TITLE
Keep Module Help

### DISCUSSION THREAD

**Discussion by TravisHaussler** [synthetichealth/synthea#1477] [General]
I am having trouble getting a Keep Module to generate any patients at all.  Is the [Keep Patients Module](https://github.com/synthetichealth/synthea/wiki/Keep-Patients-Module) generally working?
I am simply trying to select any patient that gets pregnant, using the following keep_module and command line:

``
Java Command:  java  -jar  synthea-with-dependencies.jar  -c  synthea_settings  -p  3  -g  F  -d  StudyOverrides/  -k  StudyOverrides/keep_module.json  -a  15-50  California  Los Angeles
``

**StudyOverrides/keep_module.json:**

{\
  "name": "keep",\
  "remarks": [\
    "A module that only keeps patients who give birth"\
  ],\
  "states": {\
    "Initial": {\
      "type": "Initial",\
      "conditional_transition": [\
        {\
          "transition": "Keep",\
          "condition": {\
            "condition_type": "Attribute",\
            "attribute": "pregnant",\
            "operator": "==",\
            "value": true\
          }\
        },\
        {\
          "transition": "Terminal"\
        }\
      ]\
    },\
    "Terminal": {\
      "type": "Terminal"\
    },\
    "Keep": {\
      "type": "Terminal"\
    }\
  },\
  "gmf_version": 2\
}

result:
``
java.lang.RuntimeException: Failed to produce a matching patient after 1000 attempts. Ensure that it is possible for all requested demographics to meet the criteria. (e.g., make sure there is no age restriction that conflicts with a requested condition, such as limiting age to 0-18 and requiring all patients have a condition that only onsets after 55.) If you are confident that the constraints are possible to satisfy but rare, consider increasing the value in config setting `generate.max_attempts_to_keep_patient`
``

I tried following these examples:
[Examples](https://github.com/synthetichealth/synthea/tree/0d87247879c6bd6a1bedb52784b39b611bf9fe2a/src/main/resources/keep_modules)

Additional Note:  I am using full history
``
exporter.years_of_history = 0
``

**Comment by jawalonoski** [comment]
`pregnant = true` is going to export _"women who are actively pregnant at the exact time the simulation ends."_

It is not for anyone who was ever pregnant.

To only export mothers (regardless of when the child was born), try this:

```json
{
  "name": "Mothers",
  "remarks": [
    "Keep only women who ever had children."
  ],
  "states": {
    "Initial": {
      "type": "Initial",
      "conditional_transition": [
        {
          "transition": "Keep",
          "condition": {
            "condition_type": "And",
            "conditions": [
              {
                "condition_type": "Attribute",
                "attribute": "number_of_children",
                "operator": "is not nil"
              },
              {
                "condition_type": "Attribute",
                "attribute": "number_of_children",
                "operator": ">=",
                "value": 1
              }
            ]
          }
        },
        {
          "transition": "Terminal"
        }
      ]
    },
    "Terminal": {
      "type": "Terminal"
    },
    "Keep": {
      "type": "Terminal"
    }
  },
  "gmf_version": 2
}
```

**Reply by jawalonoski** [reply]
If your goal really is women who are actively pregnant at the time the simulation runs, you could try to remove the 1000 attempt limit:

```properties
# if criteria are provided, (for example, only_dead_patients, only_alive_patients, or a "patient keep module" with -k flag)
# this is the maximum number of times synthea will loop over a single slot attempting to produce a matching patient.
# after this many failed attempts, it will throw an exception.
# set this to 0 to allow for unlimited attempts (but be aware of the possibility that it will never complete!)
generate.max_attempts_to_keep_patient = 1000
```

**Comment by TravisHaussler** [comment]
Thanks!  I am trying the code for the number_of_children as you have written, this should be good for my use case.  I couldn't get it to work at first, but when I dropped the "-a  15-50" requirement it worked, albeit a little differently than I had expected.  It failed to generate one of my 3 patients.  Does that mean that in 1000 attempts it did not generate one single person who gave birth?  That seems incredibly unlikely.

**Reply by jawalonoski** [reply]
I wouldn't have recommended dropping the age restriction, because ages outside that scope are not eligible to be pregnant according to the modules -- so if it randomly picks an ineligible age it will never succeed.

But assuming that didn't happen, then yes, that seems unlikely.

Another alternative -- modify the pregnancy (and/or related) modules so that all eligible women become pregnant within 9 months of 26-Jun-2024 (or whenever your end date is). That would certainly make it work for generating a very specific cohort.

You can either do this manually with the module builder, or try out the [module overrides](https://github.com/synthetichealth/synthea/wiki/Module-Overrides) capability.

**Comment by TravisHaussler** [comment]
Interesting.  Ok, I have perhaps honed in on what might have been going wrong when I was originally trying different keep modules.

I am using a module override for pregnancy.json.  All it does is add some Long Acting Reversible Contraceptive (LARC) procedures at the end, directly postpartum.

If I run synthea with the module override (-d) _without_ the keep_module (-k) it works
If I run synthea _without_ the module override with the keep_module it works
but if I try to run with both -d and -k, I fail to keep any patients.

Is it possible the module overrides mess up the keep module logic somehow or should I keep investigating to see if something else is happening?

(I see this warning on the keep module page: "Note also that this uses a reduced version of the module framework to run through the module. The "keep module" should primarily include Initial, Terminal, and Simple states, with transitions using whatever logic is needed to select patients. Use of other state types may produce unexpected results.") so maybe the module overrides are preventing the pregnancy module from being in this "reduced" version at all - perhaps overrides could be added to the reduced version without too much trouble?)

**Reply by jawalonoski** [reply]
The "reduced version" only applies to the keep module.

The overrides apply to all the modules, except the keep module.

There should be no conflict between the two.

`-d` and `--module_override` are **NOT** the same option.

`-d` is used to indicate the path to an alternative set of modules.

`--module_override` is used to specify a properties file that can be used to override probability distributions.

**Comment by TravisHaussler** [comment]
I am using the -d, not the other thing you mention.  do the modules from the localModuleDir (-d option) get loaded and used in the "reduced version" that the keep module uses though?

I am afraid that because my localModuleDir contains a "pregnancy.json" (my updated version for the additional LARC procedures) that the normal pregnancy.json from the jar is not loaded (which is generally good, since my intention is to replace it) and then maybe no pregnancy module is used for testing the keep_module in the "reduced version"



I am not great with Java, but from the Java I think I can see the People being created for the CriteriaCheck with these lines from /src/main/java/org/mitre/synthea/engine/Generator.java:

this.modulePredicate = getModulePredicate();
...
public Person createPerson(long personSeed, Map<String, Object> demoAttributes) {
    ...
    person.currentModules = Module.getModules(modulePredicate);


    this.modulePredicate = getModulePredicate();


But I am not sure what the values here are for getModulePredicate.  Perhaps thats causing my new pregnancy module to not be loaded for the CriteriaCheck?

private Predicate<String> getModulePredicate() {\
  if (options.enabledModules == null) {\
    return path -> true;\
  }\
  FilenameFilter filenameFilter = new WildcardFileFilter(options.enabledModules,\
    IOCase.INSENSITIVE);\
  return path -> filenameFilter.accept(null, path);\
}\

**Reply by jawalonoski** [reply]
What is supposed to happen:

- Only the keep module uses the "reduced version."
- The keep module loading is totally separate from the other models. You shouldn't mix them together (conceptually or in the same folder in your filesystem).
- If you use `-d`, Synthea still loads all the built in models, but also supplements them with the models found on the path you've specified.

If you are wondering what modules are loaded, look at the output and it will tell you:

# Normal Run
```
% ./run_synthea 

> Task :run
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
Scanned 84 modules and 151 submodules.  <---------------------------------------------- notice this
Loading submodule modules/allergies/allergy_panel.json
Loading submodule modules/allergies/drug_allergy_incidence.json
Loading submodule modules/allergies/environmental_allergy_incidence.json
Loading submodule modules/allergies/food_allergy_incidence.json
... and so forth ...
```
# Using `-d` Option

```
% ./run_synthea -d alt

> Task :run
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
Scanned 84 modules and 151 submodules.   <---------------------------------------------- notice this
Scanned 1 local modules and 0 local submodules.  <-------------------------------------- and this
Loading submodule modules/allergies/allergy_panel.json
Loading submodule modules/allergies/drug_allergy_incidence.json
Loading submodule modules/allergies/environmental_allergy_incidence.json
Loading submodule modules/allergies/food_allergy_incidence.json
... and so forth ...
Loading module modules/pregnancy.json. <------------------------------------------------ notice this
Loading module /[path]/alt/pregnancy_local.json  <-------------------------------------- and this
... and so forth ... 
```

In the above case, both pregnancy models are running simultaneously...

If you want to _replace_ the pregnancy module, the easiest method would be to delete or overwrite the version in `src/main/resources/modules`.

**Reply by TravisHaussler** [reply]
When I do not use -d I see:
...
Loading module modules/pregnancy.json
...

When I do use -d StudyOverrides/LARC I see:
...
Loading module /[path]/StudyOverrides/LARC/pregnancy.json
_and specifically no "Loading module modules/pregnancy.json" anywhere_ 
...

This is surely because they are the same file name, so the later doesn't get loaded.

My thinking is that somehow the keep_module uses neither pregnancy module when I attempt to use  "-d StudyOverrides/LARC" but I cannot prove it with the JAVA

What I can say though is if I directly add my additional changes to modules/pregnancy.json I can run it with the keep_module, so there is definitely something in the interplay here between -d and -k I feel

**Comment by TravisHaussler** [comment]
Well it looks like it was specifically breaking because I was storing my keep_module and module overrides in the same folder.  Both are valid JSON "modules" I guess, so when I called the -d option, it was loading both pregnancy.json and keep_module.json as modules, when it needed to just be loading pregnancy.json and then later finding the keep_module in a separate directory to be loaded specially.  Sorry about the deeper confusions on my part.  Perhaps a quick note about this in the Keep-Patients-Module wiki could prevent future mistakes like mine (however unlikely it may be I guess)
